src/wp-admin/includes/ajax-actions.php

```diff
@@ -4588,6 +4588,39 @@ function wp_ajax_activate_plugin() {
 		wp_send_json_error( $status );
 	}
 
+	/**
+	 * Filters the configuration data for the plugin.
+	 *
+	 * The dynamic portion of the hook name refers to the plugin slug.
+	 *
+	 * How WordPress uses this data may change in the future.
+	 *
+	 * Currently, the user will be redirected to the configuration URL
+	 * after activating the plugin.
+	 *
+	 * @since 6.5.4
+	 *
+	 * @param array {
+	 *      An array of configuration data for the plugin.
+	 *
+	 *      @type string $url The URL to configure the plugin.
+	 *                        Must be an absolute URL beginning with the administration URL.
+	 *                        Default empty string for no configuration URL.
+	 * }
+	 */
+	$configuration_data = apply_filters( 'plugin_configuration_data_' . $status['slug'], array( 'url' => '' ) );
+	if ( is_array( $configuration_data ) ) {
+		$status['configurationData'] = array();
+
+		if ( isset( $configuration_data['url'] ) ) {
+			$configuration_url = trim( $configuration_data['url'] );
+
+			if ( str_starts_with( $configuration_url, admin_url() ) ) {
```

Comment:
It may be worth changing `admin_url()` to `network_admin_url()`, which falls back to `admin_url()` anyway if it isn't a multisite. What do you think?

Reply:
Maybe? @JJJ what do you think?

Reply:
Not @JJJ but this makes sense.

Reply:
I'm not confident this approach will work for plugins that could be activated under any of the site/network/user admin area contexts (_pretend for now I didn't just say per-user plugin activations_ 😅).

BuddyPress, for example, can be active on a single site, many individual sites, or the entire network, resulting in a different URL being passed into this filter based on where activation happened.

In a multisite installation:

- If I activate BuddyPress on a single site, using network_admin_url() would fail.
- If I activate BuddyPress on a network, using admin_url() would fail.

If we are going to ask plugins to pass in their own URL, why aren't we just trusting their values? 

If we are trying to avoid plugins doing unsafe redirects here, they can still do them a million other ways (and they could've always done them previously; any evidence of this as a problem?)

Does this URL need to be validated at all, and if so, would a home_url() check sufficient (to make it context-agnostic?)

Reply:
What if ...

Instead of validating, it uses `sanitize_url()` for sanitizing "for .. redirect usage".

It then goes into `esc_url()` which cleans it with some validation (e.g. protocol).

This approach:
* is used in the same file in `wp_ajax_send_link_to_editor()` and `_wp_ajax_delete_comment_response()`.
* removes the `trim` and `str_starts_with( $configuration_url, admin_url() )` check to simplify the code.

How about this?

```php
if ( isset( $configuration_data['url'] ) && is_string( $configuration_data['url'] ) ) {
	$configuration_url = sanitize_url( $configuration_data['url'] );

	if ( ! $configuration_url ) {
		$status['configurationData']['url'] = $configuration_url;
	}
}
```

`sanitize_url()` will also return a `string`. Thus, the check before assignment skips if an empty string is returned.

Reply:
Is more validation needed such as ensuring it's in the admin?

Looking at the same file where URL is passed to the AJAX, there's no additional validation.

Looking at `wp_redirect()` and `wp_safe_redirect()`, there's no additional validation of filtered URLs.

Thinking no additional validation is needed.

Reply:
Applied the `sanitize_url()` approach in https://github.com/WordPress/wordpress-develop/pull/6606/commits/86cb5912f2510557d514851835f70d468732dacc.

Added a defensive guard to ensure the URL is a string (i.e. for better PHP 8 compat) as `sanitize_url()` requires a `string` https://github.com/WordPress/wordpress-develop/pull/6606/commits/3a4e54aafdaaa25eb876a650b68dd7809ff40658.

Reply:
Should the redirect destination be passed through `wp_sanitize_redirect()` and `wp_validate_redirect()`? If it were a PHP redirect the practice would be to use `wp_safe_redirect()`.

As the code comes from a plugin, the developer can pretty much do anything but this seems like a case in which we can protect core from coding errors.